src/libs/actions/Welcome.js

```diff
@@ -8,17 +8,60 @@ import * as Policy from './Policy';
 import ONYXKEYS from '../../ONYXKEYS';
 import NameValuePair from './NameValuePair';
 import CONST from '../../CONST';
+import createOnReadyTask from '../createOnReadyTask';
+
+const [isReady, setReady] = createOnReadyTask();
+
+let isFirstTimeNewExpensifyUser;
+let isLoadingReportData = true;
+let isLoadingPolicyData = true;
+
+/**
+ * Check that a few requests have completed so that the welcome action can proceed:
+ *
+ * - Whether we are a first time new expensify user
+ * - Whether we have loaded all policies the server knows about
+ * - Whether we have loaded all reports the server knows about
+ */
+function checkOnReady() {
+    if (!_.isBoolean(isFirstTimeNewExpensifyUser) || isLoadingPolicyData || isLoadingReportData) {
+        return;
+    }
+
+    setReady();
+}
 
-/* Flag for new users used to show welcome actions on first load */
-let isFirstTimeNewExpensifyUser = false;
 Onyx.connect({
     key: ONYXKEYS.NVP_IS_FIRST_TIME_NEW_EXPENSIFY_USER,
-    callback: val => isFirstTimeNewExpensifyUser = val,
+    initWithStoredValues: false,
+    callback: (val) => {
+        isFirstTimeNewExpensifyUser = val;
+        checkOnReady();
+    },
+});
+
+Onyx.connect({
+    key: ONYXKEYS.IS_LOADING_REPORT_DATA,
+    initWithStoredValues: false,
+    callback: (val) => {
+        isLoadingReportData = val;
+        checkOnReady();
+    },
+});
+
+Onyx.connect({
+    key: ONYXKEYS.IS_LOADING_POLICY_DATA,
+    initWithStoredValues: false,
+    callback: (val) => {
+        isLoadingPolicyData = val;
+        checkOnReady();
+    },
```

Comment:
Not a big deal, but after `setReady` we no longer need to be connected to these Onyx keys

Reply:
Actually we might need - for when you logout and login again